crates/sqlite/src/lib.rs

```diff
@@ -0,0 +1,175 @@
+                        .get(&database)
+                        .ok_or(sqlite::Error::NoSuchDatabase)?
+                        .get_connection()?,
+                )
+                .map_err(|()| {
+                    todo!("Create an error for when we reach capacity on number of databases")
+                })
+        }
+        .await)
+    }
+
+    async fn execute(
+        &mut self,
+        connection: sqlite::Connection,
+        statement: String,
+        parameters: Vec<sqlite::Value>,
+    ) -> anyhow::Result<Result<(), sqlite::Error>> {
+        Ok(async move {
+            let conn = self.get_connection(connection)?;
+            let mut statement = conn
+                .prepare_cached(&statement)
+                .map_err(|e| sqlite::Error::Io(e.to_string()))?;
+            statement
+                .execute(rusqlite::params_from_iter(convert_data(
+                    parameters.into_iter(),
+                )))
+                .map_err(|e| sqlite::Error::Io(e.to_string()))?;
+            Ok(())
+        }
+        .await)
+    }
+
+    async fn query(
+        &mut self,
+        connection: sqlite::Connection,
+        query: String,
+        parameters: Vec<sqlite::Value>,
+    ) -> anyhow::Result<Result<Vec<sqlite::Row>, sqlite::Error>> {
+        Ok(async move {
+            let conn = self.get_connection(connection)?;
+            let mut statement = conn
+                .prepare_cached(&query)
+                .map_err(|e| sqlite::Error::Io(e.to_string()))?;
+            let rows = statement
+                .query_map(
+                    rusqlite::params_from_iter(convert_data(parameters.into_iter())),
+                    |row| {
+                        let mut values = vec![];
+                        for column in 0.. {
+                            let name = row.as_ref().column_name(column);
+                            if let Err(rusqlite::Error::InvalidColumnIndex(_)) = name {
+                                break;
+                            }
+                            let name = name?.to_string();
+                            let value = row.get::<usize, ValueWrapper>(column);
+                            if let Err(rusqlite::Error::InvalidColumnIndex(_)) = value {
+                                break;
+                            }
+                            let value = value?.0;
+                            values.push(sqlite::ColumnValue { name, value });
+                        }
+                        Ok(sqlite::Row { values })
+                    },
+                )
+                .map_err(|e| sqlite::Error::Io(e.to_string()))?;
+            Ok(rows
+                .into_iter()
+                .map(|r| r.map_err(|e| sqlite::Error::Io(e.to_string())))
+                .collect::<Result<_, sqlite::Error>>()?)
+        }
+        .await)
+    }
+
+    async fn close(&mut self, connection: sqlite::Connection) -> anyhow::Result<()> {
+        Ok(async {
```

Comment:
Can replace the `async { ... }.await` with a normal block here.

Reply:
And likewise for the other methods in this `impl` block.

Reply:
Also, can we wrap these in `tokio::task::block_in_place` like we do in https://github.com/fermyon/spin/blob/b1ac4d8307948bbcd833fc689be3471f8fa174a0/crates/key-value-sqlite/src/lib.rs#L71-L138?

Reply:
Yes though we might move to a different sqlite library which has an async interface.

Reply:
So the reason I wrapped these in `async` blocks is the same reason it's done in `key-value` crate. We want the `?` to propagate the inner error of the method's return type. `async` blocks have different scoping rules than regular blocks and allow for a "try block" like control flow. I agree it's a bit ugly, but the alternatives are also not that great...

Reply:
We could do something like this: https://github.com/fermyon/spin/compare/sqlite-host-component...block-in-place

Reply:
I definitely like the `block-in-place` option until we're ready to move to an async sqlite library (which will likely do something similar under the hood).